crates/full-node/sov-db/src/state_db_nomt.rs

```diff
@@ -19,142 +20,121 @@ const BOTH: &str = "user_and_kernel_state";
+                        "Detected in-progress commit. Rolling back kernel DB."
                     );
+
+                    self.kernel.rollback(1)?;
                 }
-                kernel
-                    .rollback(1)
-                    .context("Failed to rollback kernel DB after in-progress commit detected")?;
-                commit_flag
-                    .write_status(CommitStatus::Completed)
-                    .with_context(|| {
-                        commit_flag.log_reset_instruction();
-                        "Failed to write `COMPLETED` status after rollback. Manual intervention required."
-                    })?;
-                let rolled_back_root_hash = kernel.root().into_inner();
-                tracing::info!(
-                    from_root_hash = hex::encode(current_kernel_root_hash),
-                    to_root_hash = hex::encode(rolled_back_root_hash),
-                    "Kernel namespace rollback completed"
-                );
             }
-            CommitStatus::Completed => {
-                tracing::trace!("Commit flag is `Completed`, proceeding as usual");
+            CommitStatus::CommittingUserNomt(saved_hash) => {
+                let current_user_root_hash = self.user.root().into_inner();
+
+                // User & Kernel commit was successful but later commits failed. We rollback both.
+                if saved_hash != current_user_root_hash {
+                    tracing::warn!(
+                        flag_user_root_hash = hex::encode(saved_hash),
+                        db_user_root_hash = hex::encode(current_user_root_hash),
+                        ?commit_status,
+                        "Detected in-progress commit. Rolling back kernel DB."
+                    );
+                    self.kernel.rollback(1)?;
+                    self.user.rollback(1)?;
+                } else
+                // Only Kernel commit was successful. We rollback only the kernel.
+                {
+                    tracing::warn!(
+                        ?commit_status,
+                        "Detected in-progress commit. Rolling back kernel & user DBs."
+                    );
+                    self.kernel.rollback(1)?;
+                }
             }
-        }
 
-        let user = {
-            let opts = config.get_user_options();
-            Nomt::<BinaryHasher<H>>::open(opts)?
-        };
+            CommitStatus::CommittingArchivalUserAndKernel
+            | CommitStatus::CommittingLiveUserAndKernel => {
+                tracing::warn!(
+                    ?commit_status,
+                    "Detected in-progress commit. Rolling back kernel & user DBs."
+                );
+                // User & Kernel commit was sucefull but we don't see `Success`. We rollback both User & Kernel.
+                self.kernel.rollback(1)?;
+                self.user.rollback(1)?;
+            }
+            CommitStatus::Success => {}
+        }
```

Comment:
Do we need to update flag to success at the end, so it is consistent on the next restart?